lxd/device/cdi/configure.go

```diff
@@ -0,0 +1,292 @@
+//     In our case, these edits will be interpreted either as disk or unix char mounts passed to the container.
+//     The hooks will be centralized in a single resource file that will be read and executed as a LXC `lxc.hook.mount` hook,
+//     through LXD's `callhook` command.
+//  2. We first process the 'specific device' configuration: we convert this information into a map of devices
+//     (keyed by their path given in the spec, it mapped to a map of device properties). We also collect the specific mounts (but we do not process them yet) and hooks.
+//  3. We then process the 'general device' configuration in the same fashion.
+//  4. Now we process all the mounts we collected from the spec in order to turn them into disk devices.
+//     This operations generate a side effect: it generates a list of indirect symlinks (see `specMountToNativeDev`)
+//  5. Merge all the hooks (direct + indirect) into a single list of hooks.
+func GenerateFromCDI(inst instance.Instance, cdiID ID, l logger.Logger) (*ConfigDevices, *Hooks, error) {
+	// 1. Generate the CDI specification
+	spec, err := generateSpec(cdiID, inst, l)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed to generate CDI spec: %w", err)
+	}
+
+	var targetName string
+	cdiName := cdiID.Name
+	if len(cdiName) == 1 {
+		targetName = cdiName[0]
+	} else if len(cdiName) == 2 && cdiID.Class == MIG {
+		targetName = fmt.Sprintf("%s:%s", cdiName[0], cdiName[1])
+	} else {
+		return nil, nil, fmt.Errorf("Invalid CDI locator %v", cdiName)
+	}
+
+	// Initialize the hooks as empty
+	hooks := &Hooks{ContainerRootFS: inst.RootfsPath()}
+	mounts := make([]*specs.Mount, 0)
+	configDevices := &ConfigDevices{UnixCharDevs: make(map[string]map[string]string), DiskDevs: make(map[string]map[string]string)}
+
+	// 2. Process the specific device configuration
+	for _, specDevice := range spec.Devices {
+		if specDevice.Name == targetName {
+			for _, d := range specDevice.ContainerEdits.DeviceNodes {
+				err = specDevToNativeDev(configDevices, d)
+				if err != nil {
+					return nil, nil, err
+				}
+			}
+
+			for _, hook := range specDevice.ContainerEdits.Hooks {
+				err = specHookToLXDCDIHook(hook, hooks)
+				if err != nil {
+					return nil, nil, err
+				}
+			}
+
+			mounts = append(mounts, specDevice.ContainerEdits.Mounts...)
+			break
+		}
+	}
+
+	// 3. Process the general device nodes
+	for _, generalDeviceNode := range spec.ContainerEdits.DeviceNodes {
+		err = specDevToNativeDev(configDevices, generalDeviceNode)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	// 3. Process the general hooks
+	for _, generalHook := range spec.ContainerEdits.Hooks {
+		err = specHookToLXDCDIHook(generalHook, hooks)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	// 3. Gather the general mounts
+	mounts = append(mounts, spec.ContainerEdits.Mounts...)
+
+	// CDI Mounts are converted to LXD `disk` devices
+	// If the instance is privileged, then we shift the disk device
```

Comment:
Please add a "why" component to this comment

Reply:
fixed